### PR TITLE
feat(mls): integrate CoreCrypto framework FS-646

### PR DESCRIPTION
**Pull request by David-Henner**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-646" title="FS-646" target="_blank">TaskFS-646</a>  [iOS] Integrate CoreCrypto framework
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR introduces and implements protocols and data structures that will help setting up the `CoreCrypto` framework

- `CoreCryptoConfiguration` represents the configuration required by `CoreCrypto` to be instantianted
- `CoreCryptoSetupDelegate` may be used to tell the UI that we're ready to generate the configuration. (once the client id is available)
- `CoreCryptoConfigurationProvider` generates the `CoreCryptoConfiguration` object
- `CoreCryptoProvider` is the interface used to save / retrieve a `CoreCyptoProtocol`-conforming object

### Dependencies

Needs releases with:

- [ ] https://github.com/wireapp/wire-ios-data-model/pull/1376

### Testing

- [ ] TBA

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Source/UserSession/ZMUserSession+CoreCrypto.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    var coreCrypto: CoreCryptoProtocol? {
+        get {
+            return managedObjectContext.coreCrypto
+        }
+        set {
+            managedObjectContext.coreCrypto = newValue
+        }
+    }
+}
```
Comment (johnxnguyen) [inline Source/UserSession/ZMUserSession+CoreCrypto.swift:55]: We might need to store this on the sync context as I'm guessing that's where we'll need to use it most. What do you think?

File: Source/SessionManager/SessionManager.swift [hunk 2]
```diff
@@ -1090,6 +1091,7 @@ extension SessionManager: UserSessionSelfUserClientDelegate {
             }
         }
 
+        coreCryptoSetupDelegate?.setUpCoreCryptoIfNeeded()
```
Comment (johnxnguyen) [inline Source/SessionManager/SessionManager.swift:1094]: We might be able to get away with (at least for the time being), setting up core crypto just once in the ui project when we transition to the authenticated app state. That would mean we wouldn't need this delegate at all.

File: Source/UserSession/ZMUserSession+CoreCrypto.swift [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+        do {
+            let key = try CoreCryptoKeyProvider.coreCryptoKey()
+            return CoreCryptoConfiguration(
+                path: mlsDirectory.path,
+                key: key.base64EncodedString(),
+                clientId: clientId
+            )
+        } catch {
+            fatalError(String(describing: error))
```
Comment (johnxnguyen) [inline Source/UserSession/ZMUserSession+CoreCrypto.swift:43]: It'd be good to have a TODO here for error handling.

File: Source/UserSession/ZMUserSession+CoreCrypto.swift (hunk 1 above)
Comment (David-Henner) [reply Source/UserSession/ZMUserSession+CoreCrypto.swift:55]: I agree

**Review by ghuMustafa — APPROVED** [review]

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**